src/pallene/gc.lua

```diff
@@ -28,77 +29,234 @@ local types = require "pallene.types"
+            local previous_val = input[v]
+            local new_val = true
+            input[v] = new_val
+            in_changed = in_changed or (previous_val ~= new_val)
+        end
+
+        for v, _ in pairs(input) do
+            if not output[v] and not gen[v] then
+                input[v] = nil
+                in_changed = true
             end
         end
+
+        return in_changed
     end
 
-    -- 2) Find which variables are live at each GC spot in the program.
-
-    local live_gc_vars = {} -- { cmd => {var_id}? }
-    for i, cmd in ipairs(flat_cmds) do
-        local tag = cmd._tag
-        if
-            tag == "ir.Cmd.CallStatic" or
-            tag == "ir.Cmd.CallDyn" or
-            tag == "ir.Cmd.CheckGC"
-        then
-            live_gc_vars[cmd] = {}
-            for _, v_id in ipairs(defined_variables) do
-                local a = first_definition[v_id]
-                local b = last_use[v_id]
-                if a and b and a < i and i <= b then
-                    table.insert(live_gc_vars[cmd], v_id)
-                end
+    local function merge_live(input, output)
+        for v, _ in pairs(input) do
+            output[v] = true
+        end
+    end
+
+    local function empty_set(S)
+        for v,_ in pairs(S) do
+            S[v] = nil
+        end
+    end
+
+    local succ_list = ir.get_successor_list(block_list)
+    local pred_list = ir.get_predecessor_list(block_list)
+    local block_order = ir.get_predecessor_depth_search_topological_sort(pred_list)
+
+    local dirty_flag = {} -- { block_id -> bool? } keeps track of modified blocks
+    for i = 1, #block_list do
+        dirty_flag[i] = true
+    end
+
+    local function update_block(block_i)
+        local block_succs = succ_list[block_i]
+        local block_preds = pred_list[block_i]
+        local state = state_list[block_i]
+
+        -- last block's output is supposed to be fixed
+        if block_i ~= #block_list then
+            empty_set(state.output)
+            for _,succ in ipairs(block_succs) do
+                local succ_in = state_list[succ].input
+                merge_live(succ_in, state.output)
+            end
+        end
+
+        local in_changed = apply_gen_kill_sets(state)
+        if in_changed then
+            for _, pred in ipairs(block_preds) do
+                dirty_flag[pred] = true
             end
         end
     end
 
-    local variable_is_live_at_gc = {}  -- { var_id => boolean }
-    for v_id = 1, #func.vars do
-        variable_is_live_at_gc[v_id] = false
+    repeat
+        local found_dirty_block = false
+        for _,block_i in ipairs(block_order) do
+            if dirty_flag[block_i] then
+                found_dirty_block = true
+                -- CAREFUL: we have to clean the dirty flag BEFORE updating the block or else we
+                -- will do the wrong thing for auto-referencing blocks
+                dirty_flag[block_i] = nil
```

Comment:
Use `=false` here to avoid putting holes in the table. Holes can't be iterated with ipairs, and can cause Lua to reallocate the table.

Use `nil` for sparse sets only.

Reply:
Didn't know about the reallocation thing. I'll fix this.